### PR TITLE
mload mstore and mstore8 opcode modification

### PR DESCRIPTION
**Pull request by ashWhiteHat**
Part of https://github.com/privacy-scaling-explorations/zkevm-specs/issues/230

I introduced `StepState` and `Instruction` to `MLOAD`, `MSTORE` and `MSTORE8` opcodes.
The circuit constraints and logics are aligned with zkevm-circuits.
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/71ef2647c79044654d57dd17bab4285088bc0989/zkevm-circuits/src/evm_circuit/execution/memory.rs#L39

I would appreciate it if you could confirm.
Thank you!

### CODE REVIEW

File: src/zkevm_specs/evm/execution/memory.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        for idx in range(32):
+            instruction.memory_lookup(
+                RW.Write if is_store == FQ(1) else RW.Read, address.expr() + idx, src_id
+            )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(34 - (is_mstore8 * 31)),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(is_store * -2),
```
Comment (han0110) [inline src/zkevm_specs/evm/execution/memory.py:38]: Should increase the `stack_pointer` if it's store (pop 2 elements)?

```suggestion
        stack_pointer=Transition.delta(is_store * 2),
```

File: tests/evm/test_memory.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        rw_table=rw_dictionary.rws,
+    )
+
+    address = offset + 1 + (is_not_mstore8 * 31)
+    next_mem_size, memory_gas_cost = memory_expansion(curr_memory_size, address)
+    gas = Opcode.MLOAD.constant_gas_cost() + memory_gas_cost
+
+    rw_counter = 35 - (is_mstore8 * 31)
+    program_counter = 66 - (is_mload * 33)
+    stack_pointer = 1020 + (is_mload * 2)
```
Comment (han0110) [inline tests/evm/test_memory.py:115]:
```suggestion
    stack_pointer = 1022 + (is_store * 2)
```

**Review by han0110 — COMMENTED** [review]
Overall LGMT. Nice refactor! Just got a small question about the `stack_pointer`

**Comment by ashWhiteHat** [comment]
Hi @han0110 
Thank you for the review!
It was weird as you mentioned.
I modified according to your review.

I would appreciate it if you could confirm.
Thank you!

**Review by han0110 — APPROVED** [review]
LGTM now!

File: src/zkevm_specs/evm/execution/memory.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    value = instruction.stack_push() if is_mload == FQ(1) else instruction.stack_pop()
+
+    src_id = instruction.call_context_lookup(CallContextFieldTag.TxId)
+    memory_offset = instruction.curr.memory_size
+    next_memory_size, memory_expansion_gas_cost = instruction.memory_expansion(
+        memory_offset, address.expr() + FQ(1) + (is_not_mstore8 * FQ(31))
+    )
+
+    if is_mstore8 == FQ(1):
+        instruction.memory_lookup(RW.Write, address.expr())
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/memory.py:26]:
```suggestion
        instruction.is_equal(instruction.memory_lookup(RW.Write, address.expr()), value.le_bytes[0])
```
Add equality check to value with the lookup

File: src/zkevm_specs/evm/execution/memory.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    if is_not_mstore8 == FQ(1):
+        for idx in range(32):
+            instruction.memory_lookup(
+                RW.Write if is_store == FQ(1) else RW.Read, address.expr() + idx, src_id
+            )
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/memory.py:32]:
```suggestion
            instruction.is_equal(instruction.memory_lookup(
                RW.Write if is_store == FQ(1) else RW.Read, address.expr() + idx, src_id
            ), value.le_bytes[31-idx])
```
Add equality check to value with the lookup

File: src/zkevm_specs/evm/execution/memory.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    address = instruction.stack_pop()
+
+    is_mload = instruction.is_equal(opcode, Opcode.MLOAD)
+    is_mstore8 = instruction.is_equal(opcode, Opcode.MSTORE8)
+    is_store = FQ(1) - is_mload
+    is_not_mstore8 = FQ(1) - is_mstore8
+
+    value = instruction.stack_push() if is_mload == FQ(1) else instruction.stack_pop()
+
+    src_id = instruction.call_context_lookup(CallContextFieldTag.TxId)
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/memory.py:19]: I didn't find this lookup in the circuit.  I see that you use it in the `memory_lookup`, but that function takes a `call_id` and not a `tx_id`.  So I think this should be removed, and the `memory_lookup` for `is_not_mstore8` should get the 3rd parameter removed (the `src_id`).

**Review by ed255 — CHANGES_REQUESTED** [review]

**Comment by ashWhiteHat** [comment]
Hi @ed255
Thank you for the review.
Your suggestion makes sense to me and I changed PR accordingly.

I would appreciate it if you could confirm.
Thank you!

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for working on this update!

**Pull request merged**